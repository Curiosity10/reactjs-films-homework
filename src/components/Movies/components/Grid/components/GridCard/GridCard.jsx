import React from 'react';
import PropTypes from 'prop-types';
import styles from './GridCard.scss';
import { OverlayCard, DefaultCard } from './components';

const GridCard = ({ movie, isActive, changeActiveCard }) => {
  const {
    rating,
    title,
    genres,
    overview,
    image,
    id,
  } = movie;
  const maxLength = 215;
  const truncate = description => (
    description.length > maxLength ? `${description.slice(0, maxLength)}...` : description
  );
  return (
    <article className={styles.card}>
      <img height={isActive ? '75%' : '100%'} src={image} alt="Poster" />
      { isActive
        ? (
          <DefaultCard
            id={id}
            title={title}
            rating={rating}
            changeActiveCard={changeActiveCard}
            genres={genres}
          />
        )
        : (
          <OverlayCard
            overview={overview}
            truncate={truncate}
            title={title}
            rating={rating}
            changeActiveCard={changeActiveCard}
            genres={genres}
          />
        )}
    </article>
  );
};

GridCard.propTypes = {
  movie: PropTypes.shape({
    rating: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    genres: PropTypes.array.isRequired,
    overview: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired,
  }).isRequired,
  isActive: PropTypes.bool.isRequired,
  changeActiveCard: PropTypes.func.isRequired,
};

export default GridCard;
