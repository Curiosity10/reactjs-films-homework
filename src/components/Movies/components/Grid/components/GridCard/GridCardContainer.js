import { connect } from 'react-redux';
import GridCard from './GridCard';
import { fetchVideoKey } from '../../../../../../modules/app/appActions';

export default connect(null, { fetchVideoKey })(GridCard);
