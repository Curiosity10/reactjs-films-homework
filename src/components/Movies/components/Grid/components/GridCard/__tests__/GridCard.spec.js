import React, { useState } from 'react';
import { create, act } from 'react-test-renderer';
import GridCard from '../GridCard';

describe('GridCard component renders correctly', () => {
  const movie = {
    title: 'test',
    id: 1,
    genre_ids: [
      {
        id: 12,
        name: 'Action',
      },
      {
        id: 16,
        name: 'Adventure',
      },
    ],
    overview: 'desc',
  };
  const activeCard = 1;
  const mockFunction = jest.fn();

  it('GridCard renders and works correctly', () => {
    const tree = create(
      <GridCard
        movie={movie}
        activeCard={activeCard}
        changeActiveCard={mockFunction}
        isActive={false}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Set overlay works correctly', () => {
    const Wrapper = () => {
      const film = {
        title: 'test',
        id: 2,
        genre_ids: [
          {
            id: 12,
            name: 'Action',
          },
          {
            id: 16,
            name: 'Adventure',
          },
        ],
        overview: 'desc',
      };
      const [currentCard, setCurrentCard] = useState(1);
      return (
        <>
          <GridCard
            movie={film}
            changeActiveCard={id => setCurrentCard(id)}
            isActive={film.id !== currentCard}
          />
        </>
      );
    };
    const tree = create(
      <Wrapper />,
    );

    const watchBtn = tree.root.findByProps({ 'aria-label': 'Watch' });
    const mockWatchBtn = jest.fn(watchBtn.props.onClick);
    const button = tree.root.findByProps({ 'aria-label': 'View Info' });

    act(() => {
      mockWatchBtn();
      button.props.onClick();
    });

    const overlay = tree.toJSON().children[1];
    expect(mockWatchBtn).toHaveBeenCalled();
    expect(overlay.props).toEqual({ className: 'overlay' });
  });

  it('Set default card with short overview works correctly', () => {
    const Wrapper = () => {
      const film = {
        title: 'test',
        id: 2,
        genre_ids: [
          {
            id: 12,
            name: 'Action',
          },
          {
            id: 16,
            name: 'Adventure',
          },
        ],
        overview: 'desc',
      };
      const [currentCard, setCurrentCard] = useState(2);
      return (
        <>
          <GridCard
            movie={film}
            changeActiveCard={id => setCurrentCard(id)}
            isActive={film.id !== currentCard}
          />
        </>
      );
    };

    const tree = create(
      <Wrapper />,
    );

    const watchBtn = tree.root.findByProps({ 'aria-label': 'Watch Now' });
    const mockWatchBtn = jest.fn(watchBtn.props.onClick);
    const button = tree.root.findByProps({ 'aria-label': 'Close' });

    act(() => {
      mockWatchBtn();
      button.props.onClick();
    });

    const onHover = tree.toJSON().children[1];
    expect(onHover.props).toEqual({ className: 'onHover' });
    expect(mockWatchBtn).toHaveBeenCalled();
  });
});
