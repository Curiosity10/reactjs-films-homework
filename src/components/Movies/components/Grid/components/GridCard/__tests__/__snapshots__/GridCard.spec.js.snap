// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridCard component renders correctly GridCard renders and works correctly 1`] = `
<article
  className="card"
>
  <img
    alt="Poster"
    height="100%"
  />
  <div
    className="overlay"
  >
    <button
      aria-label="Close"
      className="close"
      onClick={[Function]}
      type="button"
    />
    <div
      className="wrapper"
    >
      <div>
        <h3
          className="title"
          title="test"
        >
          <a
            href="/movie/:1"
            onClick={[Function]}
          >
            test
          </a>
        </h3>
      </div>
    </div>
    <p
      className="description"
    >
      desc
    </p>
    <button
      aria-label="Watch Now"
      className="btn primary"
      onClick={[Function]}
      type="button"
    >
      Watch Now
    </button>
  </div>
</article>
`;
